cultcargo/pfb-clean.yml

```diff
@@ -0,0 +1,32 @@
+_include:
+  - genesis/cult-cargo-base.yml
+
+cabs:
+  _include: genesis/pfb-clean/latest/uncabbedcabs.yml
+
+  pfb_init:
+    image: ${vars.cult-cargo.registry}/pfb-clean:cc${vars.cult-cargo.version}
```

Comment:
This is my old ugly way... the image entry has since been changed to a structure (see quartical below for an example). So the right way is

```yml
image:
      _use: vars.cult-cargo.images
      name: pfb-clean
```

Or, if you want to save repetition, define your own vars entry right in this file, and reuse it:

```yml
vars:
  cult-cargo:
    pfb-clean:
      image:
        _use: vars.cult-cargo.images
        name: pfb-clean

cabs:
  ...
  pfb_init:
    image:
      _use: vars.cult-cargo.pfb-clean.image
```

Reply:
Done. Can't remember the reason for using _ in the first place. Something I am not sure about is why the substitution of _ with - is not working for parameters. For example, say I define the following 
```yaml
pfb-init:
  command: pfb.workers.init.init
  flavour: python
  policies:
    pass_missing_as_none: true
    prefix: --
    replace: {_: '-'}

  _include:
    - init.yaml
```
If I inspect the docs for the cab I get
```bash
(ccargo) ╭─landman@LAPTOP-ED4N50Q2 ~/software/cult-cargo  ‹qc*›
╰─➤  stimela -C doc cultcargo/pfb-clean.yml pfb-init
2024-03-04 09:52:35 STIMELA INFO: starting
2024-03-04 09:52:35 STIMELA INFO: clearing 2 cached config(s) from cache
2024-03-04 09:52:35 STIMELA INFO: loading configuration
2024-03-04 09:52:35 STIMELA INFO: loading config from
/home/landman/venvs/ccargo/lib/python3.10/site-packages/stimela/stimela.conf
2024-03-04 09:52:35 STIMELA INFO: loaded config from
/home/landman/venvs/ccargo/lib/python3.10/site-packages/stimela/stimela.conf
2024-03-04 09:52:35 STIMELA INFO: saving config dependencies to ./stimela.config.deps
2024-03-04 09:52:35 STIMELA INFO: will load recipe/config file 'cultcargo/pfb-clean.yml'
⠦  stimela ⠦  starting 0:00:02  stimela doc cultcargo/pfb-clean.yml pfb-init
└── Cab: pfb-init
    ├── command: pfb.workers.init.init
    ├── image: quay.io/stimela2/pfb-clean:cc0.1.2
    ├── Required inputs:
    │   └── output_filename  str        Basename of output
    │       log_directory    str        Directory to write logs and performance reports to.
    │       ms               List[str]  Path to measurement set
    └── Optional inputs:
        └── host_address            str                  Address where the distributed client lives. Uses LocalCluster
                                                         if no address is provided and scheduler is set to distributed.
            nworkers                int                  Number of worker processes. Use with distributed scheduler.
                                                         [default: 1]
            nthreads_dask           int                  Number of dask threads. This will largely determine the memory
                                                         footprint. Used as nthreads per worker when using distributed
                                                         scheduler. [default: 1]
            nvthreads               int                  Number of threads used to scale vertically (eg. for FFTs and
                                                         gridding). Each dask thread can in principle spawn this many
                                                         threads. Will attempt to use half the available threads by
                                                         default.
            scheduler               str                  Which dask scheduler to use. [default: threads, choices:
                                                         threads, single-threaded, sync, distributed, processes]
            product                 str                  Imaging products to produce. Options are I, Q, U, V. Only
                                                         single Stokes products are currently supported [default: I]
            scans                   List[int]            List of SCAN_NUMBERS to image. Defaults to all. Input as string
                                                         eg. '[0,2]' if running from CLI
            ddids                   List[int]            List of DATA_DESC_ID's to images. Defaults to all. Input as
                                                         string eg. '[0,1]' if running from CLI
            fields                  List[int]            List of FIELD_ID's to image. Defaults to all. Input as string
                                                         eg. '[0,1,2]' if running from CLI
            freq_range              str                  Frequency range to image in Hz. Specify as a string with colon
                                                         delimiter eg. '1e9:1.1e9'
            overwrite               bool                 Allow overwrite of output xds [default: False]
            radec                   Optional[List[any]]  Rephase all images to this radec specified in radians
            data_column             str                  Data column to image. Must be the same across MSs [default:
                                                         DATA]
            weight_column           str                  Column containing natural weights. Must be the same across MSs
            sigma_column            str                  Column containing standard devations. Will be used to
                                                         initialise natural weights if detected. Must be the same across
                                                         MSs
            flag_column             str                  Column containing data flags. Must be the same across MSs
                                                         [default: FLAG]
            gain_table              List[str]            Path to Quartical gain table containing NET gains. There must
                                                         be a table for each MS and glob(ms) and glob(gt) should match
                                                         up
            gain_term               str                  Which gain term to use. Default is NET [default: NET]
            integrations_per_image  int                  Number of time integrations per image. Default (-1, 0, None) ->
                                                         dataset per scan. [default: -1]
            channels_per_image      int                  Number of channels per image. Default (-1, 0, None) -> dataset
                                                         per spw. [default: -1]
            precision               str                  Gridding precision [default: double, choices: single, double]
            bda_decorr              float                BDA decorrelation factor. Only has an effect if less than one
                                                         [default: 1.0]
            beam_model              str                  Beam model to use
            max_field_of_view       float                Field of view in degrees [default: 3.0]
            chan_average            int                  Average this number if channels together [default: 1]
```
so the replacement didn't take place. What gives?

Reply:
@landmanbester policies only relate to how parameters are converted into command-line options. In your case you shouldn't even be using a policies section, since Stimela is calling Python directly. Policies.replace is meant for third-party executables.

For calling Python code, there's anyway an implicit replacement of "-" in parameter names by "_".

I gather that you would prefer to have your parameter names use dashes -- just go ahead and change them all, it should just work.

``prefix: --`` is also unnecessary (again, this is for calling third-party tools). ``clickify_parameters`` uses a ``--`` prefix by default anyway when it constructs your CLI. 

 

Also, given that you're S3-compatible, I suggest using ``List[URI]`` for the MS input.

Reply:
Also @landmanbester, ``- init.yaml`` should be ``- (.)init.yaml`` -- remember https://github.com/caracal-pipeline/stimela/issues/198.

Reply:
Ok, I see. I have long forgotten why I put those options in there in the first place. I've changed it but now I have run into my request limit so will have to wait an hour before updating.